### PR TITLE
[not for merge] Trinitus01 changes to amazonechocontrol

### PR DESCRIPTION
**Pull request by J-N-K**

### CODE REVIEW

**Comment by J-N-K** [comment]
@Trinitus01 Doesn't look that bad. Some general comments: 

- The `.properties` file has wrong encoding, ISO-8859-1 is needed here, it looks like you used UTF-8
- `@author`-tag needs a real name, for a new file like `@author John Doe - Initial Contribution
- Some formatting looks wrong. Please apply spotless-formatter (`mvn spotless:apply -pl bundles\org.openhab.binding.amazonechocontrol`) 

I guess most conflicts should be solved then.

**Comment by Trinitus01** [comment]
Thanks for the feedback i will take a look into it.
I also forgot to add the "Signed-off-by:" to the commit (else i read it is necessary), will this be a problem?
So maybe i have to do a commit amend and a push force or do the fork stuff again :(

I realised that the new sound channel is not necessary because someone on the community wrote he implemented it via SSML but in the new version it doesn't work anymore. I also tried to do some days ago and thought it is not possible there is a difference between speak and sound. Now i know it is/was possible and i would check this too.

**Comment by J-N-K** [comment]
We can fix the sign-off later, no need to worry.

**Comment by Trinitus01** [comment]
Today i found a solution for parallel tts (the known way is sequenced) so now it is possible to send one tts to all echos with only one request (i thought this is only possible via announcement, like michi also thought i think). A sequenced tts arrives the echos step by step and produce a resound effect, very bad, but a parallel tts is like you send x requests to every single echo at the same time, but in one request. This also enables ssml audio aso. The best thing is, for the user there is no change in usage anymore. (Because of the version major changes aso) The user can use announcement like before and tts like before without the too many request issue. There was also an issue with the web socket because if you send a parallel tts, a lot of web socket messages produces a lot of requests again. I fixed it, so at the moment it works very well. Currently i am testing all the stuff. After testing, i will fix the issues you reported and do a push. Hopefully tomorrow or maybe the day after.

**Comment by J-N-K** [comment]
Unfortunately I had to do some refactoring to remove potential bugs. We‘ll see if we can simply merge or need some more detailed work.

**Comment by Trinitus01** [comment]
Maybe you can help me, i have the following problem: a method is called many times in one second and i want to queue them and call them with a delay of x seconds. The first call should run the code immediately and the second, third, aso should called delayed. If the delay runs out, it should call immediately again. Any idea how to implement this. In real life it means for example: 3 commands with the same text, lets say "hello" are send to different echo items and 1 command with another text lets say "hello too" is send to the different echo items too (1 second next to the first one). For now only "hello too" is spoken because the second request overlap the first one. A delay is useless because each sentence has another length. So the only way is to send the tts, receive the tts is arrived via websocket and unlock the tts for the next request: method1 send tts, lock and wait for method2, method2 wait for receiving tts via websocket and unlock method1, now !method1! can call the second tts, and so on. I will find a solution for sure but maybe you have an idea and save some time^^

**Comment by Trinitus01** [comment]
Wrong desciption correction:  method1 send tts, lock and wait for method2, method2 wait for receiving tts via websocket and unlock method1, now !method1! can call the second tts, and so on. I will find a solution for sure but maybe you have an idea and save some time^^

**Comment by J-N-K** [comment]
Something like this:

```

    private LinkedBlockingQueue<String> messageQueue = new LinkedBlockingQueue<>();
    private AtomicBoolean messageQueueRunning = new AtomicBoolean();
    private ScheduledFuture<?> messageSenderUnblockFuture;

    // call this to send a message
    public void queueMessage(String message) {
        messageQueue.add(message);
        if (messageQueueRunning.compareAndSet(false,true)) {
            processQueue();
        }
    }

    // called by the websocket when the response is received or after 30s by the scheduled unblocker
    public void messageSent() {
        ScheduledFuture<?> messageSenderUnblockFuture = this.messageSenderUnblockFuture;
        if (messageSenderUnblockFuture != null) {
            messageSenderUnblockFuture.cancel(true);
            this.messageSenderUnblockFuture = null;
        }
        if (!messageQueue.isEmpty()) {
            processQueue();
        } else {
            messageQueueRunning.set(false);
        }
    }

    private void processQueue() {
        String nextMessage = messageQueue.poll();

        // do whatever is needed to send the message

        // set a timeout, if we do not receive the websocket response within 30s, unblock the queue
        messageSenderUnblockFuture = scheduler.schedule(this::messageSent, 30, TimeUnit.SECONDS);
    }
```

**Comment by Trinitus01** [comment]
Yes, your example looks very near to what i am looking for! I will try to implement it and give a feedback. Very big thanks!

**Comment by Trinitus01** [comment]
Your example work like it should and its a very nice to handle a sync process, i will take this as a big lesson. But, sadly, amazon send the web socket as soon as the request arrives. So method1 call a tts and method2 immediately receives that the tts is called and calls method1 again. (like it should) But the time between the method1 calls are too short and this produces a "too many"-ex. It seems there is no other solution than set a time waiting for the next request. But the time has to be longer then the speaking time. Else it ends in something like: Sentence1 "This is a test and the test takes a lot of time", Sentence2 "And this is another test", Result "This ist a test and the...is another test" (Overlap). If it is not possible to recognise if a tts is finished, there is no possibility to start the next one after the first one :( In my case i use a tts to say: "xyz is in front of the house door" and if the door opens "the front door is going to open". But they overlap and i only receive one or a half of them. One idea is to count the words and set 1 second for every word, but this is -for sure- bumbling.

**Comment by J-N-K** [comment]
```
 private LinkedBlockingQueue<String> messageQueue = new LinkedBlockingQueue<>();
    private AtomicBoolean messageQueueRunning = new AtomicBoolean();
    private ScheduledFuture<?> messageSenderUnblockFuture;

    // call this to send a message
    public void queueMessage(String message) {
        messageQueue.add(message);
        if (messageQueueRunning.compareAndSet(false,true)) {
            processQueue();
        }
    }

    private void processQueue() {
        if (!messageQueue.isEmpty()) {
            String nextMessage = messageQueue.poll();

            // do whatever is needed to send the message

            int wordcount = nextMessage.split("\\s+").length;

            // process next
            messageSenderUnblockFuture = scheduler.schedule(this::messageSent, wordcount, TimeUnit.SECONDS);
        } else {
            messageQueueRunning.set(false);
            messageSenderUnblockFuture = null;
        }
    }

```

Maybe like this. But don't forget to check the future in `dispose` and cancel it if non-null, otherwise we are leaking resources again.

**Comment by Trinitus01** [comment]
Hi,

with your code i was able to fix it. But i used the count of all letters in a sentence instead the words. 100 milliseconds per letter seems to be a perfect value. No overlapping anymore, nearly same delays between sentences, very nice. Should i test it for some days or should i push it so you are able to take a look at? I still have to fix the encoding and format.

Br

**Comment by Trinitus01** [comment]
Pushed.
It seems there are only some little conflicts left.

Br

**Comment by Trinitus01** [comment]
Btw there is a description in the README how to use a json for tts (called Expert) with an option sound:

{ "sound": true, "speak":"<Speak>" "title": "<Title>", "body": "<Body Text>", "volume": 20}

This "sound"-option is useless, it is only used to check if sound true and ssml speak is not allowed but thats all. the sound parameter is never used in any functional way.

Br

File: bundles/org.openhab.binding.amazonechocontrol/README.md [hunk 1]
```diff
@@ -180,6 +182,10 @@ Sample for the Thing echo1 only. But it will work in the same way for the other
 Take a look in the channel description above to know, which channels are supported by your thing type.
 
 ```
+// Account
+String Echo_Living_Room_SendMessage            "SendMessage"                           {channel="amazonechocontrol:account:account1:sendMessage"}
+String Echo_Living_Room_AccountAnnouncement    "AccountAnnouncement"                   {channel="amazonechocontrol:account:account1:accountAnnouncement"}
```
Comment (J-N-K) [inline bundles/org.openhab.binding.amazonechocontrol/README.md]: Shouldn't the cannel be named `sendAnnouncement` to keep it consistent with the `sendMessage`?

**Review by J-N-K — COMMENTED** [review]
There are some more comments. I wouldlike to address them by creating a PR to your repo. Can you enable that?

Am I right that this is now fully backward compatible?

**Comment by Trinitus01** [comment]
Yes it should be fully backward compatible. But keep in mind that i dont have that much use cases like other people. On the base of code i see no difference in usage.

I will change accountAnnouncement to sendAnnouncement tomorrow in the morning.

How to enable "it"? Maybe u can tell me else i will google it tomorrow in the morning, too.

Br

**Comment by Trinitus01** [comment]
Found an issue :( It seems textToSpeechQueueRunning (the AtomicBoolean) stay true and queuedTextToSpeech is never called again (tts stop working). So i still have to take a closer look at it.

Edit: Not reproducible anymore, happens twice, so still keeping a look on it.
Edit: Found the reason, an unhandled exception, so still keeping a look on it.

**Comment by Trinitus01** [comment]
I added a "Branch Protection Rule" to * Branches with the options "Require pull request reviews before merging" and "Require review from Code Owners". Was it the correct thing to enable?

Br

**Comment by Trinitus01** [comment]
Hi,

sorry to ask you for help again :(

I queued tts and web socket but i realised the solution is not to queue the individual functions.
The real major problem is the function "makeRequest" in the "Connection" class.

Every feature like tts, announcement, web socket, aso use this single function to send a request, for sure.
The knowledge is to queue and delay !this! single function and not the functions for tts, ann...aso.
To solve the problem for tts and websocket i used your example, but its not possible for the "makeRequest" function, because it returns a value and is not a void function.
Now, i really don't know how to queue and delay the "makeRequest" function, i tried and i failed.

btw this is the big hole of "too many requests"-issue, every thing else was only a "hole"-fix for the single functions. My current version is stable but its not perfect because the major fix is missing. The structure of the binding in this case is not well, hopefully you can give me a hint to solve this problem.

Edit: changed makeRequest to synchronized to reduce to problem. Tried to add a sleep but this is not the right way, so removed it. For now it is stable else it is not the best way to handle all the stuff.

Br

**Pull request closed without merging**